sqlmesh/core/engine_adapter/mssql.py

```diff
@@ -187,10 +187,12 @@ def query_factory() -> Query:
             # pymssql doesn't convert Pandas Timestamp (datetime64) types
             # - this code is based on snowflake adapter implementation
             for column, kind in columns_to_types_create.items():
+                # pymssql errors if the column contains a datetime.date object
+                if kind.is_type("date"):  # type: ignore
+                    df[column] = pd.to_datetime(df[column]).dt.strftime("%Y-%m-%d")  # type: ignore
+
                 if is_datetime64_any_dtype(df.dtypes[column]):
-                    if kind.is_type("date"):  # type: ignore
-                        df[column] = pd.to_datetime(df[column]).dt.strftime("%Y-%m-%d")  # type: ignore
-                    elif getattr(df.dtypes[column], "tz", None) is not None:  # type: ignore
+                    if getattr(df.dtypes[column], "tz", None) is not None:  # type: ignore
                         # MSSQL requires a colon in the offset (+00:00) so we use isoformat() instead of strftime()
```

Comment:
Just FMI: if we previously hit the `is_type("date")`, then it seems like we wouldn't hit the the timezone `elif` branch below it. Is this still true now that the former check has been moved before the `datetime64_any_dtype` branch? i.e. will we ever hit both the `is_type("date")` branch _and_ the timezone one now? Does that matter?

Reply:
Good point - once the first conversion is done the column is guaranteed not to be datetime64 so we can skip that evaluation.

If the column type is date then any time-related info was just being ignored during the cast to date on the mssql side, so not hitting that branch wouldn't be an issue anyway.